Sources/NIOHPACK/HPACKEncoder.swift

```diff
@@ -0,0 +1,299 @@
+        var indexing: HPACKIndexing
+    }
+    
+    private enum EncoderState {
+        case idle(maxTableSize: Int)
+        case resized(currentMaxTableSize: Int, smallestMaxTableSize: Int?)
+        case encoding(maxTableSize: Int)
+    }
+    
+    // private but tests
+    var headerIndexTable: IndexedHeaderTable
+    
+    private var state: EncoderState
+    private var buffer: ByteBuffer!
+    
+    /// Whether to use Huffman encoding.
+    public let useHuffmanEncoding: Bool
+    
+    /// The current size of the dynamic table.
+    ///
+    /// This is defined as the sum of [name] + [value] + 32 for each header.
+    public var dynamicTableSize: Int {
+        return self.headerIndexTable.dynamicTableLength
+    }
+    
+    /// The current maximum size to which the dynamic header table may grow.
+    public private(set) var allowedDynamicTableSize: Int {
+        get { return self.headerIndexTable.dynamicTableAllowedLength }
+        set { self.headerIndexTable.dynamicTableAllowedLength = newValue }
+    }
+    
+    /// The hard maximum size of the dynamic header table, set via an HTTP/2
+    /// SETTINGS frame.
+    public var maximumDynamicTableSize: Int {
+        get { return self.headerIndexTable.maxDynamicTableLength }
+        set { self.headerIndexTable.maxDynamicTableLength = newValue }
+    }
+    
+    /// Sets the maximum size for the dynamic table and encodes the new value
+    /// at the start of the current packed header block to send to the peer.
+    ///
+    /// - note: The size cannot grow beyond the maximum allotted by the most recent
+    ///         acknowledged SETTINGS_HEADER_TABLE_SIZE. It will truncate to that
+    ///         value automatically.
+    ///
+    /// - Parameter size: The new maximum size for the dynamic header table.
+    public mutating func setDynamicTableSize(_ size: Int) throws {
+        // clip to the largest permitted size
+        let size = min(size, self.maximumDynamicTableSize)
+        
+        switch self.state {
+        case .idle:
+            self.state = .resized(currentMaxTableSize: size, smallestMaxTableSize: nil)
+        case let .resized(currentSize, smallestSize):
+            if let smallest = smallestSize {
+                if size < smallest {
+                    self.state = .resized(currentMaxTableSize: size, smallestMaxTableSize: nil)
+                } else {
+                    self.state = .resized(currentMaxTableSize: size, smallestMaxTableSize: smallest)
+                }
+            } else if size < currentSize {
+                self.state = .resized(currentMaxTableSize: size, smallestMaxTableSize: nil)
+            } else {
+                self.state = .resized(currentMaxTableSize: size, smallestMaxTableSize: currentSize)
+            }
+        case .encoding:
+            throw NIOHPACKErrors.EncoderAlreadyActive()
+        }
+        
+        // set the new size on our dynamic table
+        self.allowedDynamicTableSize = size
```

Comment:
If we are always setting `currentMaxTableSize` on our dynamic table (which I think is the right thing to do), then we don't need to carry a duplicate copy of it in our `EncoderState` enum. We can remove the first `Int` from each case, and save ourselves 8 bytes.

Reply:
That's what I'd thought, but then I thought you knew something I didn't 😄